4337/test/utils/webauthn.ts

```diff
@@ -0,0 +1,244 @@
+      response: {
+        clientDataJSON: ethers.toUtf8Bytes(JSON.stringify(clientData)).buffer,
+        attestationObject: b2ab(CBOR.encode(attestationObject)),
+      },
+      type: 'public-key',
+    }
+  }
+
+  /**
+   * This is a shim for `navigator.credentials.get` method.
+   * See <https://w3c.github.io/webappsec-credential-management/#dom-credentialscontainer-get>.
+   *
+   * @param options The public key credential request options.
+   * @returns A public key credential with an assertion response.
+   */
+  get({ publicKey }: CredentialRequestOptions): PublicKeyCredential<AuthenticatorAssertionResponse> {
+    const credential = publicKey.allowCredentials
+      .flatMap(({ id }) => this.#credentials.filter((c) => c.rp === publicKey.rpId && c.id === ethers.hexlify(id)))
+      .at(0)
+    if (credential === undefined) {
+      throw new Error('credential not found')
+    }
+
+    // <https://w3c.github.io/webauthn/#dictionary-client-data>
+    const clientData = {
+      type: 'webauthn.get',
+      challenge: base64UrlEncode(publicKey.challenge).replace(/=*$/, ''),
+      origin: `https://${publicKey.rpId}`,
+    }
+
+    // <https://w3c.github.io/webauthn/#sctn-authenticator-data>
+    // Note that we use a constant 0 value for signCount to simplify things:
+    // > If the authenticator does not implement a signature counter, let the signature counter
+    // > value remain constant at zero.
+    const authenticatorData = ethers.solidityPacked(
+      ['bytes32', 'uint8', 'uint32'],
+      [
+        ethers.sha256(ethers.toUtf8Bytes(publicKey.rpId)),
+        0x01, // flags = user_present
+        0, // signCount
+      ],
+    )
+
+    // <https://w3c.github.io/webauthn/#sctn-op-get-assertion>
+    // <https://w3c.github.io/webauthn/#fig-signature>
+    const signature = p256.sign(
+      ethers.getBytes(ethers.concat([authenticatorData, ethers.sha256(ethers.toUtf8Bytes(JSON.stringify(clientData)))])),
+      credential.pk,
+      {
+        lowS: false,
+        prehash: true,
+      },
+    )
+
+    return {
+      id: base64UrlEncode(credential.id),
+      rawId: ethers.getBytes(credential.id),
+      response: {
+        clientDataJSON: ethers.toUtf8Bytes(JSON.stringify(clientData)).buffer,
+        authenticatorData: ethers.getBytes(authenticatorData).buffer,
+        signature: signature.toDERRawBytes(false).buffer,
+        userHandle: credential.user,
+      },
+      type: 'public-key',
+    }
+  }
+}
+
+/**
+ * Encode bytes using the Base64 URL encoding.
+ *
+ * See <https://www.rfc-editor.org/rfc/rfc4648#section-5>
+ *
+ * @param data data to encode to `base64url`
+ * @returns the `base64url` encoded data as a string.
+ */
+export function base64UrlEncode(data: BytesLike | ArrayBufferLike): string {
+  const bytes = isArrayBuffer(data) ? new Uint8Array(data) : ethers.getBytes(data)
+  return Buffer.from(bytes).toString('base64url')
+}
+
+function b2ab(buf: Buffer): ArrayBuffer {
```

Comment:
Imo it's better to stick to Uint8Array here (buffer extends Uint8Array)

Fun read: https://sindresorhus.com/blog/goodbye-nodejs-buffer

Reply:
`Buffer` be gone!

acf4235e71d0535c3b0b8cddeb4bb16ec62aff7b